### PR TITLE
Fix Db::executeQuery() for null parameter

### PR DESCRIPTION
**Pull request by W0rma**
Seems like https://github.com/Codeception/module-db/pull/48 broke the following code

```php
$I->seeInDatabase('some_table', ['some_column => null]);
```
After updating `codeception/module-db` from `3.1.0` to `3.1.1` I get the following error:

```
[TypeError] Codeception\Lib\Driver\Db::isBinary(): Argument #1 ($string) must be of type string, null given, called in /var/www/html/vendor/codeception/module-db/src/Codeception/Lib/Driver/Db.php on line 297
```

This PR should fix that by using `PDO::PARAM_NULL` if `null` was provided in the criteria.

Fixes https://github.com/Codeception/module-db/issues/64

### CODE REVIEW

**Review by Arhell — APPROVED** [review]

**Comment by pascal08** [comment]
Experiencing the same problem. Downgrading to v3.0.1 helps, but introduces [another problem](https://github.com/Codeception/module-db/pull/45) fixed in v3.1.0.

**Comment by tecbird** [comment]
same here, please merge

**Comment by troy-rudolph** [comment]
We have the same problem.  #48 causes many of our tests to fail as show below.  Please merge and ship this fix.
```
00:38:41  [37;41;1m  [TypeError] Codeception\Lib\Driver\Db::isBinary(): Argument #1 ($string) must be of type string, null given, called in /web/manager-jenkins/jenkins-Manager-PR-17842-5/www/vendor/codeception/module-db/src/Codeception/Lib/Driver/Db.php on line 297  [39;49;22m
00:38:41  [37;41;1m                                                                                                                                                                                                                                                         [39;49;22m
00:38:41  [33m
00:38:41  Scenario Steps:
00:38:41  [39m
00:38:41   6. $I->updateInDatabase("Tests",{"Publis...},{"TestID...}) at [32mtests/api/observations/ObservationsFiltersCest.php:222[39m
00:38:41   5. // I am going to update the Status for getting archived and abandoned Observations (in Tests table)
00:38:41   4. $I->grabFromDatabase("Tests","TestID",{"CourseID":40}) at [32mtests/api/observations/ObservationsFiltersCest.php:215[39m
00:38:41   3. $I->grabFromDatabase("Tests","TestID",{"CourseID":38}) at [32mtests/api/observations/ObservationsFiltersCest.php:210[39m
00:38:41   2. $I->useDataGeneratorTool({"actions":[{"action":"tun...}) at [32mtests/api/observations/ObservationsFiltersCest.php:194[39m
00:38:41   1. $I->getMicroStamp() at [32mtests/api/observations/ObservationsFiltersCest.php:17[39m
00:38:41  
00:38:41  #1  Codeception\Module\Db->updateInDatabase
00:38:41  #2  /web/manager-jenkins/jenkins-Manager-PR-17842-5/www/tests/codeception2/tests/_support/_generated/ApiTesterActions.php:4686
00:38:41  #3  /web/manager-jenkins/jenkins-Manager-PR-17842-5/www/tests/codeception2/tests/api/observations/ObservationsFiltersCest.php:222
00:38:41  #4  SM\ObservationsFiltersCest->_before
00:38:41  #5  /web/manager-jenkins/jenkins-Manager-PR-17842-5/www/vendor/bin/codecept:119
00:38:41  [33mArtifacts:[39m
```

**Review by troy-rudolph — APPROVED** [review]

**Comment by MattiaPeiretti** [comment]
Can someone please merge @W0rma's PR? Real companies are relying on this and the fact that a fix is present for 5 days and it's just waiting to be merged is quite sad.

File: src/Codeception/Lib/Driver/Db.php
Comment (guraurobi) [inline src/Codeception/Lib/Driver/Db.php:299]: is_string($param) is really not needed here, as the method mb_detect_encoding() is always gonna cast the first parameter to (string). 

Maybe the method isBinary() should be completly replaced with its contents, because currently that method does not detect a binary string or anything binary related.

I suggest using `!mb_detect_encoding($string, null, true)` directly here.

File: composer.json
Comment (guraurobi) [inline composer.json:20]: Not needed.

**Review by guraurobi — CHANGES_REQUESTED** [review]

File: composer.json
Comment (troy-rudolph) [reply composer.json:20]: We have  ext-mbstring installed.  It still fails.

Comment (troy-rudolph) [reply composer.json:20]: Please merge this.

File: src/Codeception/Lib/Driver/Db.php
Comment (troy-rudolph) [reply src/Codeception/Lib/Driver/Db.php:299]: In our case, the arg is null because our test is trying to find rows where a column has a null value.  Please merget his.

**Comment by laoneo** [comment]
Please merge

File: src/Codeception/Lib/Driver/Db.php
Comment (eerison) [reply src/Codeception/Lib/Driver/Db.php:299]: @troy-rudolph 

did you try to use the suggestion mentioned by gurau?

File: composer.json
Comment (eerison) [reply composer.json:20]: maybe it should be added in your composer.json

**Comment by MattiaPeiretti** [comment]
33 days old. 
<div><br/>via <a href="https://giphy.com/hero0fwar/">hero0fwar</a> on <a href="https://giphy.com/gifs/hero0fwar-sad-pablo-escobar-jc2PkKKr3clTBekMzn">GIPHY</a></div>

Still waiting.

**Comment by eerison** [comment]
> 33 days old.
> 
> 
> Still waiting. I believe in you, you can do it!

the quickly solution is, @troy-rudolph try to apply @guraurobi suggestions, otherwise it will be in discussion, for a long time, because it still have open discussion, and they aren't hard to solve.

File: src/Codeception/Lib/Driver/Db.php
Comment (Daredzik) [reply src/Codeception/Lib/Driver/Db.php:299]: @eerison I tried, solution suggested from @guraurobi is incorrect: 
`[TypeError] mb_detect_encoding(): Argument #1 ($string) must be of type string, null given`
`is_string` do the trick here

**Comment by Daredzik** [comment]
Merge this, suggestion to use `mb_detect_encoding` is wrong:
`[TypeError] mb_detect_encoding(): Argument #1 ($string) must be of type string, null given`
`is_string` do the trick here

**Comment by Daredzik** [comment]
@Naktibalda or @sergeyklay can you check this one?

**Comment by eerison** [comment]
> @Naktibalda or @sergeyklay can you check this one?

Just to point out, this issue was introduced on `3.1.1`, you can back one version and wait until this PR be merged.

```json
  "conflict": {
    "symfony/symfony": "*",
    "codeception/module-db": ">3.1.0"
  },

```

File: src/Codeception/Lib/Driver/Db.php
Comment (eerison) [reply src/Codeception/Lib/Driver/Db.php:299]: Yep that is true `mb_detect_encoding` requires a string.

> as the method mb_detect_encoding() is always gonna cast the first parameter to (string).

@guraurobi it gonna cast if you are not using `declare(strict_types=1)`, otherwise it will throw the exception mentioned by @Daredzik .

> Maybe the method isBinary() should be completly replaced with its contents, because currently that method does not detect a binary string or anything binary related.

IMO it is not related with this issue, i guess we could implement this in other PR, I don't see any issue with the current code, and as we can see it was already tested by a lot of guys 😄.

Then I guess we could mark this thread as solved, don't you @guraurobi ?

Btw: @Daredzik could you confirm if others suggestions from @guraurobi worked or not?

File: composer.json
Comment (eerison) [reply composer.json:20]: @W0rma could you remove this line please, I also think it works without it.

Comment (Daredzik) [reply composer.json:20]: @eerison it should work without it.

Comment (eerison) [reply composer.json:20]: then removing it, IMO all threads are solved

Comment (W0rma) [reply composer.json:20]: Isn't `mb_detect_encoding()` only available if the `mbstring` extension is installed?
That's the reason why I added this requirement.

Btw. codeception itself has such requirement as well ( https://github.com/Codeception/Codeception/blob/5.0/composer.json#L21 ).

Comment (eerison) [reply composer.json:20]: I don't know,  but isBinary was already using it for months and no one complains about it. And it is not explicitly in php docs 

https://www.php.net/manual/en/function.mb-detect-encoding.php

And have this requirement in codeception doesn't mean it is necessary in this lib.

Comment (Daredzik) [reply composer.json:20]: @W0rma  @eerison 
Yes, `mb_detect_encoding()` is part of the `mbstring` extension (https://www.php.net/manual/en/mbstring.installation.php). However, this module can't run without the main Codeception library, so it also uses Codeception dependencies. But we can leave it here for clarity.

**Comment by MattiaPeiretti** [comment]
42 days old. 
<div>

Still waiting. I believe in you, you can do it!

File: composer.json
Comment (eerison) [reply composer.json:20]: > Yes, mb_detect_encoding() is part of the mbstring extension

if it requires, then we should have this in composer.json

```json
"ext-mbstring": "*",
```

and @guraurobi we already `mb_detect_encoding` merged, then we should have this in composer.json, have https://github.com/Codeception/module-db/blob/185889230bccfc65d2c03624879ade5357ec43f9/src/Codeception/Lib/Driver/Db.php#L350

IMO **everything as clarified**, and the PR is **ready to merge**.

**Review by sergeyklay — APPROVED** [review]

**Comment by sergeyklay** [comment]
THank you for the patch and I'm sorry for the delay.

**Comment by MattiaPeiretti** [comment]
<div>

**Pull request merged**
